File: weblayer/browser/java/org/chromium/weblayer_private/NewTabCallbackProxy.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.weblayer_private;

import android.os.RemoteException;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.weblayer_private.interfaces.NewTabType;

/**
 * Owns the c++ NewTabCallback class, which is responsible for forwarding all
 * NewTabCallback calls to this class, which in turn forwards to ITabClient.
 */
@JNINamespace("weblayer")
public final class NewTabCallbackProxy {
    private long mNativeNewTabCallbackProxy;
    private final TabImpl mTab;

    public NewTabCallbackProxy(TabImpl tab) {
        mTab = tab;
        mNativeNewTabCallbackProxy =
                NewTabCallbackProxyJni.get().createNewTabCallbackProxy(this, tab.getNativeTab());
    }

    public void destroy() {
        NewTabCallbackProxyJni.get().deleteNewTabCallbackProxy(mNativeNewTabCallbackProxy);
        mNativeNewTabCallbackProxy = 0;
    }

    @NewTabType
    private static int implTypeToJavaType(@ImplNewTabType int type) {
        switch (type) {
            case ImplNewTabType.FOREGROUND:
                return NewTabType.FOREGROUND_TAB;
            case ImplNewTabType.BACKGROUND:
                return NewTabType.BACKGROUND_TAB;
            case ImplNewTabType.NEW_POPUP:
                return NewTabType.NEW_POPUP;
            case ImplNewTabType.NEW_WINDOW:
                return NewTabType.NEW_WINDOW;
        }
        assert false;
        return NewTabType.FOREGROUND_TAB;
    }

    @CalledByNative
    public void onNewTab(long nativeTab, @ImplNewTabType int mode) throws RemoteException {
        // This class should only be created while the tab is attached to a fragment.
        assert mTab.getBrowser() != null;
        TabImpl tab =
                new TabImpl(mTab.getProfile(), mTab.getBrowser().getWindowAndroid(), nativeTab);
        mTab.getBrowser().addTab(tab);
        mTab.getClient().onNewTab(tab.getId(), mode);
    }

    @CalledByNative
    private void onCloseTab() throws RemoteException {
        mTab.getClient().onCloseTab();
    }

    @NativeMethods
    interface Natives {
        long createNewTabCallbackProxy(NewTabCallbackProxy proxy, long tab);
        void deleteNewTabCallbackProxy(long proxy);
    }
}

File: weblayer/browser/new_tab_callback_proxy.cc
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "weblayer/browser/new_tab_callback_proxy.h"

#include "base/trace_event/trace_event.h"
#include "url/gurl.h"
#include "weblayer/browser/java/jni/NewTabCallbackProxy_jni.h"
#include "weblayer/browser/tab_impl.h"

using base::android::AttachCurrentThread;
using base::android::ScopedJavaLocalRef;

namespace weblayer {

NewTabCallbackProxy::NewTabCallbackProxy(JNIEnv* env, jobject obj, TabImpl* tab)
    : tab_(tab), java_impl_(env, obj) {
  DCHECK(!tab_->has_new_tab_delegate());
  tab_->SetNewTabDelegate(this);
}

NewTabCallbackProxy::~NewTabCallbackProxy() {
  tab_->SetNewTabDelegate(nullptr);
}

void NewTabCallbackProxy::OnNewTab(Tab* tab, NewTabType type) {
  JNIEnv* env = AttachCurrentThread();
  TRACE_EVENT0("weblayer", "Java_NewTabCallbackProxy_onNewTab");
  Java_NewTabCallbackProxy_onNewTab(env, java_impl_,
                                    static_cast<TabImpl*>(tab)->GetJavaTab(),
                                    static_cast<int>(type));
}

static jlong JNI_NewTabCallbackProxy_CreateNewTabCallbackProxy(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& proxy,
    jlong tab) {
  return reinterpret_cast<jlong>(
      new NewTabCallbackProxy(env, proxy, reinterpret_cast<TabImpl*>(tab)));
}

static void JNI_NewTabCallbackProxy_DeleteNewTabCallbackProxy(JNIEnv* env,
                                                              jlong proxy) {
  delete reinterpret_cast<NewTabCallbackProxy*>(proxy);
}

}  // namespace weblayer
